File: unixfs/rpc/client/fs-cursor-client.ts
import type { FSCursorChange } from '../../fs-cursor.js'
import { ErrReleased, UnixFSError } from '../../errors/errors.js'
import type { NodeType } from '../../block/fstree.pb.js'
import type {
  FSCursorChange as ProtoFSCursorChange,
  FSCursorClientResponse,
} from '../rpc.pb.js'
import type { FSCursorService } from '../rpc_srpc.pb.js'
import { RemoteFSCursor } from './fs-cursor-remote.js'
import { RemoteFSCursorOps } from './fs-cursor-ops-remote.js'

// FSCursorClient is the cursor client session manager.
// It manages the event watching loop from the FSCursorClient streaming RPC.
// The cursor will be released if the streaming RPC is canceled.
export class FSCursorClient {
  released = false
  readonly client: FSCursorService
  readonly clientHandleId: bigint
  rootCursor!: RemoteFSCursor
  private ctrl: AbortController
  private cursors = new Map<bigint, RemoteFSCursor>()
  private ops = new Map<bigint, RemoteFSCursorOps>()
  private cursorOps = new Map<bigint, bigint>()

  private constructor(
    client: FSCursorService,
    clientHandleId: bigint,
    ctrl: AbortController,
  ) {
    this.client = client
    this.clientHandleId = clientHandleId
    this.ctrl = ctrl
  }

  // build constructs and initializes the FSCursorClient.
  // Does not return until the init message is received from the remote.
  static async build(
    svcClient: FSCursorService,
    signal?: AbortSignal,
  ): Promise<FSCursorClient> {
    const ctrl = new AbortController()
    if (signal) {
      signal.addEventListener('abort', () => ctrl.abort(), { once: true })
    }
    const stream = svcClient.FSCursorClient({}, ctrl.signal)
    const iter = stream[Symbol.asyncIterator]()

    const first = await iter.next()
    if (first.done) {
      ctrl.abort()
      throw new Error('stream closed before init')
    }

    const msg = first.value
    if (msg.body?.case === 'unixfsError') {
      ctrl.abort()
      const err = UnixFSError.fromProto(msg.body.value)
      if (err) {
        throw err
      }
      throw new Error('unknown unixfs error in init')
    }
    if (msg.body?.case !== 'init') {
      ctrl.abort()
      throw new Error('expected init')
    }

    const init = msg.body.value
    const clientId = init.clientHandleId ?? 0n
    const rootId = init.cursorHandleId ?? 0n
    if (!clientId || !rootId) {
      ctrl.abort()
      throw new Error('empty handle IDs in init')
    }

    const fsc = new FSCursorClient(svcClient, clientId, ctrl)
    const rootCursor = new RemoteFSCursor(fsc, rootId)
    fsc.rootCursor = rootCursor
    fsc.cursors.set(rootId, rootCursor)
    fsc.execute(iter)
    return fsc
  }

  // execute is the background loop managing cursor change events.
  private async execute(
    iter: AsyncIterator<FSCursorClientResponse>,
  ): Promise<void> {
    try {
      for (;;) {
        const { done, value } = await iter.next()
        if (done) {
          break
        }
        if (value.body?.case === 'cursorChange') {
          this.handleCursorChange(value.body.value)
        }
      }
    } finally {
      this.release()
    }
  }

  // handleCursorChange handles an incoming cursor change message.
  private handleCursorChange(ch: ProtoFSCursorChange): void {
    const id = ch.cursorHandleId
    if (!id) {
      return
    }

    const cursor = this.cursors.get(id)
    if (!cursor) {
      return
    }

    if (ch.released) {
      cursor.released = true
      this.cursors.delete(id)
      const opsId = this.cursorOps.get(id)
      if (opsId !== undefined) {
        const existingOps = this.ops.get(opsId)
        if (existingOps) {
          existingOps.released = true
          this.ops.delete(opsId)
        }
        this.cursorOps.delete(id)
      }
    }

    if (cursor.cbs.length) {
      const change: FSCursorChange = {
        cursor,
        released: ch.released ?? false,
        offset: ch.offset ?? 0n,
        size: ch.size ?? 0n,
      }
      cursor.cbs = cursor.cbs.filter((cb) => cb(change))
    }
  }

  // ingestCursor creates or retrieves a RemoteFSCursor for a handle ID.
  ingestCursor(id: bigint): RemoteFSCursor {
    let c = this.cursors.get(id)
    if (!c || c.released) {
      c = new RemoteFSCursor(this, id)
      this.cursors.set(id, c)
    }
    return c
  }

  // resolveOps creates or retrieves a RemoteFSCursorOps for a handle ID.
  resolveOps(
    cursorId: bigint,
    opsId: bigint,
    nodeType: NodeType | undefined,
    name: string,
  ): RemoteFSCursorOps {
    const existing = this.ops.get(opsId)
    if (
      existing &&
      !existing.released &&
      existing.name === name &&
      existing.nodeType === nodeType
    ) {
      return existing
    }
    const oldOpsId = this.cursorOps.get(cursorId)
    if (oldOpsId !== undefined) {
      const oldOps = this.ops.get(oldOpsId)
      if (oldOps) {
        oldOps.released = true
        this.ops.delete(oldOpsId)
      }
    }
    const cursor = this.cursors.get(cursorId)
    if (!cursor) {
      throw ErrReleased
    }
    const newOps = new RemoteFSCursorOps(cursor, opsId, nodeType ?? 0, name)
    this.ops.set(opsId, newOps)
    this.cursorOps.set(cursorId, opsId)
    return newOps
  }

  // removeOps removes an ops entry from the map.
  removeOps(opsId: bigint): void {
    this.ops.delete(opsId)
  }

  // release releases the filesystem cursor client.
  release(): void {
    if (this.released) {
      return
    }
    this.released = true
    this.ctrl.abort()
    for (const op of this.ops.values()) {
      op.released = true
    }
    this.ops.clear()
    for (const cursor of this.cursors.values()) {
      cursor.released = true
    }
    this.cursors.clear()
    this.cursorOps.clear()
  }

  [Symbol.dispose](): void {
    this.release()
  }
}

File: db/unixfs/rpc/client/fs-cursor-client.go
package unixfs_rpc_client

import (
	"context"
	"errors"
	"sync"

	"github.com/s4wave/spacewave/db/unixfs"
	unixfs_rpc "github.com/s4wave/spacewave/db/unixfs/rpc"
)

// FSCursorClient is the cursor returned by GetProxyCursor on FSCursor.
// It is recommended to use FSCursor instead as it will automatically re-connect.
//
// this cursor manages the event watching loop on FSCursor.
// the cursor will be released if the FSCursorClient RPC is canceled.
type FSCursorClient struct {
	// remoteFSCursor implements FSCursor with remote rootCursorHandleID.
	*remoteFSCursor
	// ctx is the context to use for global operations (release handle)
	ctx context.Context
	// cancel cancels the context for the cursor client
	cancel context.CancelFunc
	// client is the client object for requests
	client unixfs_rpc.SRPCFSCursorServiceClient
	// clientHandleID is the client handle id to use for requests.
	clientHandleID uint64
	// mtx guards below fields
	mtx sync.Mutex
	// cursors contains the mapping from remote cursor id to cursor
	cursors map[uint64]*remoteFSCursor
	// ops contains the mapping from remote ops id to ops
	ops map[uint64]*remoteFSCursorOps
	// cursorOps contains the mapping from remote cursor id to remote ops id
	cursorOps map[uint64]uint64
}

// BuildFSCursorClient constructs & initializes the FSCursorClient, starting the
// management goroutine and initializing a client id with the remote service.
//
// does not return until the init message is received from the remote.
// the context is used for the persistent goroutine.
func BuildFSCursorClient(rctx context.Context, client unixfs_rpc.SRPCFSCursorServiceClient) (*FSCursorClient, error) {
	ctx, ctxCancel := context.WithCancel(rctx)
	strm, err := client.FSCursorClient(ctx, &unixfs_rpc.FSCursorClientRequest{})
	if err != nil {
		ctxCancel()
		return nil, err
	}

	resp, err := strm.Recv()
	if err != nil {
		ctxCancel()
		_ = strm.Close()
		return nil, err
	}

	// handle error response
	if err := resp.GetUnixfsError().ToGoError(); err != nil {
		ctxCancel()
		_ = strm.Close()
		return nil, err
	}

	// handle successful init
	initMsg := resp.GetInit()
	if initMsg == nil {
		ctxCancel()
		_ = strm.Close()
		return nil, errors.New("unexpected non-init msg as first response to FSCursorClient")
	}

	clientHandleID, rootCursorHandleID := initMsg.GetClientHandleId(), initMsg.GetCursorHandleId()
	if clientHandleID == 0 {
		ctxCancel()
		_ = strm.Close()
		return nil, errors.New("unexpected empty client handle id in fs cursor client init")
	}
	if rootCursorHandleID == 0 {
		ctxCancel()
		_ = strm.Close()
		return nil, errors.New("unexpected empty root cursor handle id in fs cursor client init")
	}

	fsc := &FSCursorClient{
		ctx:            ctx,
		cancel:         ctxCancel,
		client:         client,
		clientHandleID: clientHandleID,
		cursors:        make(map[uint64]*remoteFSCursor, 1),
		ops:            make(map[uint64]*remoteFSCursorOps),
		cursorOps:      make(map[uint64]uint64),
	}
	fsc.remoteFSCursor = newRemoteFSCursor(fsc, rootCursorHandleID)
	fsc.cursors[rootCursorHandleID] = fsc.remoteFSCursor
	go fsc.execute(ctx, strm)

	return fsc, nil
}

// execute is the goroutine managing the FSCursorClient.
func (c *FSCursorClient) execute(ctx context.Context, strm unixfs_rpc.SRPCFSCursorService_FSCursorClientClient) {
	defer func() {
		c.Release()
		c.mtx.Lock()
		for _, ops := range c.ops {
			ops.released.Store(true)
		}
		c.ops = make(map[uint64]*remoteFSCursorOps, 0)
		for _, remoteCursor := range c.cursors {
			remoteCursor.released.Store(true)
		}
		c.cursors = make(map[uint64]*remoteFSCursor, 0)
		c.cursorOps = make(map[uint64]uint64, 0)
		c.mtx.Unlock()
		_ = strm.Close()
	}()

	msg := &unixfs_rpc.FSCursorClientResponse{}
	for {
		msg.Reset()
		if err := strm.RecvTo(msg); err != nil {
			return
		}

		switch resp := msg.GetBody().(type) {
		case *unixfs_rpc.FSCursorClientResponse_CursorChange:
			if ch := resp.CursorChange; ch != nil {
				c.handleCursorChange(ctx, ch)
			}
		}
	}
}

// handleCursorChange handles an incoming cursor change message.
func (c *FSCursorClient) handleCursorChange(ctx context.Context, ch *unixfs_rpc.FSCursorChange) {
	cursorHandleID := ch.GetCursorHandleId()
	if cursorHandleID == 0 {
		return
	}

	c.mtx.Lock()
	defer c.mtx.Unlock()

	cursor, ok := c.cursors[cursorHandleID]
	if !ok {
		return
	}

	// if released, remove from the cursors set.
	if ch.Released {
		cursor.released.Store(true)
		delete(c.cursors, cursorHandleID)

		// release associated ops object, if applicable.
		opsID, opsOk := c.cursorOps[cursorHandleID]
		if opsOk {
			delete(c.cursorOps, cursorHandleID)
			ops, opsOk := c.ops[opsID]
			if opsOk {
				ops.released.Store(true)
				delete(c.ops, opsID)
			}
		}
	}

	// fire the event to any listeners
	if len(cursor.cbs) != 0 {
		changeObj := ch.ToFSCursorChange()
		for _, cb := range cursor.cbs {
			if cb != nil {
				cb(changeObj)
			}
		}
	}
}

// ingestCursorLocked ingests a FSCursor while mtx is locked.
func (c *FSCursorClient) ingestCursorLocked(cursorHandleID uint64) *remoteFSCursor {
	retCursor, retCursorOk := c.c.cursors[cursorHandleID]
	if !retCursorOk || retCursor.released.Load() {
		retCursor = newRemoteFSCursor(c.c, cursorHandleID)
		c.c.cursors[cursorHandleID] = retCursor
	}
	return retCursor
}

// Release releases the filesystem cursor client.
// All sub-cursors will be automatically canceled as well.
func (c *FSCursorClient) Release() {
	c.released.Store(true)
	c.cancel()
}

// _ is a type assertion
var _ unixfs.FSCursor = ((*FSCursor)(nil))
